includes/Domain/Utils/McpAnnotationMapper.php

```
@@ -0,0 +1,180 @@
+				$ability_annotations,
+				$mcp_field,
+				$config['ability_property']
+			);
+
+			if ( ! $value_info['has_value'] ) {
+				continue;
+			}
+
+			$normalized = self::normalize_annotation_value( $config['type'], $value_info['value'] );
+			if ( null === $normalized ) {
+				continue;
+			}
+
+			$result[ $mcp_field ] = $normalized;
+		}
+
+		return $result;
+	}
+
+	/**
+	 * Resolve the annotation value, preferring WordPress-format overrides when available.
+	 *
+	 * @param array  $annotations     Raw annotations from the ability.
+	 * @param string $mcp_field       The MCP field name.
+	 * @param string $ability_property Optional WordPress-format field name.
+	 *
+	 * @return array{has_value:bool,value:mixed}
+	 */
+	private static function resolve_annotation_value( array $annotations, string $mcp_field, string $ability_property ): array {
+		// WordPress-format overrides take precedence when present.
+		if ( '' !== $ability_property && array_key_exists( $ability_property, $annotations ) && ! is_null( $annotations[ $ability_property ] ) ) {
+			return array(
+				'has_value' => true,
+				'value'     => $annotations[ $ability_property ],
+			);
+		}
+
+		if ( array_key_exists( $mcp_field, $annotations ) && ! is_null( $annotations[ $mcp_field ] ) ) {
+			return array(
+				'has_value' => true,
+				'value'     => $annotations[ $mcp_field ],
+			);
+		}
+
+		return array(
+			'has_value' => false,
+			'value'     => null,
+		);
+	}
+
+	/**
+	 * Normalize annotation values to the types expected by MCP.
+	 *
+	 * @param string $field_type Expected MCP type (boolean, string, array, number).
+	 * @param mixed  $value      Raw annotation value.
+	 *
+	 * @return mixed|null Normalized value or null if invalid.
+	 */
+	private static function normalize_annotation_value( string $field_type, $value ) {
+		switch ( $field_type ) {
+			case 'boolean':
+				return (bool) $value;
+
+			case 'string':
+				if ( ! is_scalar( $value ) ) {
+					return null;
```

I wonder if it would make more sense to return a `WP_Error` when invalid, instead?

---

We can do that, but I do not wnat to validate at this stage. Here I'm just normalizing the data included. 
We have the `mcp_adapter_validation_enabled` setting, which enables component validation if the user wants it.

---

Hmm... then shouldn't we be including whatever the value is in the returned results? The fact that we're omitting things from the results means that we are validating — i.e. if it's not valid then we omit it.

---

I see what you mean, returning null does effectively exclude invalid values, which could be seen as a form of validation. My intent here, though, is to keep this step focused purely on normalization and avoid introducing validation logic at this stage.

I also don’t want to throw errors or risk breaking the entire MCP functionality if a tool annotation happens to be defined incorrectly. Returning null here ensures the system remains resilient and continues working even when a single annotation is malformed.

The explicit validation path is already available through the `mcp_adapter_validation_enabled` filter, which users can enable if they prefer stricter behavior.

Anyways, I'm open to discuss a  refactor related to normalization/validation functionality, but this should be done on another issue/PR as this one is getting really big :)

---

Agreed, this isn't hurting anything for now, so we can move on.

Just to get my thought out, I'm actually open to things effectively working as they do now. My only thought was to use `_doing_it_wrong()` or something like that to communicate _something_ so it doesn't simply silently fail and move along without any clue as to what happened.